File: leetcode/javascript/array/SlidingWindow/findMaxSumSubarrayLengthK(easy)/findMaxSumSubarrayLengthK.js
/* Fixed window size

Учитывая целочисленный массив nums и целое число k, найдите сумму подмассива с наибольшей суммой, длина которой равна k.
*/

/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
// time complexity: O(n), space complexity: O(1)
var findMaxSumSubarrayLengthK = function(nums, k) {
    let curr = 0; // Some data to track the window
    for (let i = 0; i < k; i++) {
        curr += nums[i];
    }

    let ans = curr;
    for (let i = k; i < nums.length; i++) {
        curr += nums[i] - nums[i - k]; // Add arr[i] to window and Remove arr[i - k] from window
        ans = Math.max(ans, curr); // Update ans
    }

    return ans;
}

console.log(findMaxSumSubarrayLengthK([3, -1, 4, 12, -8, 5, 6], 4)); // 18 <- [3, -1, 4, 12,]

File: leetcode/go/array/SlidingWindow/findMaxSumSubarrayLengthK(easy)/findMaxSumSubarrayLengthK.go
package main

import "fmt"

/* Fixed window size

Учитывая целочисленный массив nums и целое число k, найдите сумму подмассива с наибольшей суммой, длина которой равна k.
*/

func main() {
	nums := []int{3, -1, 4, 12, -8, 5, 6}
	fmt.Println(findMaxSumSubarrayLengthK(nums, 4)) // 18 <- [3, -1, 4, 12,]
}

// findMaxSumSubarrayLengthK находит сумму подмассива с наибольшей суммой, длина которой равна k.
func findMaxSumSubarrayLengthK(nums []int, k int) int {
	sum := 0 // Текущая сумма подмассива длины k

	// Вычисляем сумму первого окна длины k
	for i := 0; i < k; i++ {
		sum += nums[i]
	}

	result := sum // Инициализируем ответ суммой первого окна

	// Скользим окном по массиву
	for i := k; i < len(nums); i++ {
		sum += nums[i] - nums[i-k] // Добавляем nums[i] и удаляем nums[i-k]
		if sum > result {          // Обновляем ans, если текущая сумма больше
			result = sum
		}
	}

	return result
}
